generator/routing_index_generator.cpp

```diff
@@ -0,0 +1,145 @@
+#include "generator/routing_index_generator.hpp"
+
+#include "routing/bicycle_model.hpp"
+#include "routing/car_model.hpp"
+#include "routing/index_graph.hpp"
+#include "routing/pedestrian_model.hpp"
+#include "routing/vehicle_model.hpp"
+
+#include "indexer/feature.hpp"
+#include "indexer/feature_processor.hpp"
+#include "indexer/index.hpp"
+#include "indexer/routing_section.hpp"
+#include "indexer/scales.hpp"
+
+#include "coding/file_name_utils.hpp"
+
+#include "platform/country_file.hpp"
+#include "platform/local_country_file.hpp"
+
+#include "base/logging.hpp"
+
+#include "std/shared_ptr.hpp"
+#include "std/unique_ptr.hpp"
+#include "std/unordered_map.hpp"
+#include "std/vector.hpp"
+
+using namespace feature;
+using namespace platform;
+using namespace routing;
+
+namespace
+{
+uint32_t constexpr kFixPointFactor = 100000;
+
+inline m2::PointI PointDToPointI(m2::PointD const & p) { return m2::PointI(p * kFixPointFactor); }
+
+uint64_t CalcLocationKey(m2::PointD const & point)
+{
+  m2::PointI const pointI(PointDToPointI(point));
+  return (static_cast<uint64_t>(pointI.y) << 32) + static_cast<uint64_t>(pointI.x);
+}
+
+class Processor final
+{
+public:
+  Processor(string const & dir, string const & country)
+    : m_pedestrianModel(make_unique<PedestrianModelFactory>()->GetVehicleModelForCountry(country))
+    , m_bicycleModel(make_unique<BicycleModelFactory>()->GetVehicleModelForCountry(country))
+    , m_carModel(make_unique<CarModelFactory>()->GetVehicleModelForCountry(country))
+  {
+    LocalCountryFile localCountryFile(dir, CountryFile(country), 1 /* version */);
+    m_index.RegisterMap(localCountryFile);
+    vector<shared_ptr<MwmInfo>> info;
+    m_index.GetMwmsInfo(info);
+    CHECK_EQUAL(info.size(), 1, ());
+    CHECK(info[0], ());
+  }
+
+  void operator()(FeatureType const & f)
+  {
+    if (!IsRoad(f))
+      return;
+
+    uint32_t const id = f.GetID().m_index;
+    f.ParseGeometry(FeatureType::BEST_GEOMETRY);
+    size_t const pointsCount = f.GetPointsCount();
+    if (pointsCount == 0)
+      return;
+
+    for (size_t fromSegId = 0; fromSegId < pointsCount; ++fromSegId)
```

Comment:
> number of segments is equal to number of points minus one, right?
yes

> is it correct to add 4th segment to the 4th point?
It is correct
Joint keeps point based intersections, so i need all points, including last.

Example, 2 roads.
Road 1: p0 -- p1 -- p2 -- p3
Road 2: p0 -- p1 -- p2
Joint: [R1,p3], [R2,p1]